bundler/spec/cache/gems_spec.rb

```diff
@@ -289,22 +289,49 @@
       expect(cached_gem("rack-1.0.0")).to exist
     end
 
-    it "raises an error when the gem file is altered and produces a different checksum" do
+    it "raises an error when the gem is altered and produces a different checksum" do
+      pending "This test depends on lockfile checksums being enabled by default"
```

Comment:
I don't understand why this one won't trigger the error. I enable checksums in the lockfile.

Reply:
I think it may be because we currently don't produce the error either if the checksum section is there, but there was no locked checksum for the gem that was altered (like in the lockfile you added to the spec).

If I modify the spec like this, it passes:

```diff
diff --git a/bundler/spec/cache/gems_spec.rb b/bundler/spec/cache/gems_spec.rb
index 2f5da4e7e..bb07bfc08 100644
--- a/bundler/spec/cache/gems_spec.rb
+++ b/bundler/spec/cache/gems_spec.rb
@@ -289,15 +289,32 @@
       expect(cached_gem("rack-1.0.0")).to exist
     end
 
-    it "raises an error when the gem file is altered and produces a different checksum" do
+    it "raises an error when the gem is altered and produces a different checksum" do
       cached_gem("rack-1.0.0").rmtree
       build_gem "rack", "1.0.0", :path => bundled_app("vendor/cache")
+
+      expected_checksums = checksums_section do |c|
+        c.checksum gem_repo1, "actionpack", "2.3.2"
+        c.checksum gem_repo1, "activesupport", "2.3.2"
+        c.checksum gem_repo1, "rack", "1.0.0"
+      end
+
       simulate_new_machine
 
+      lockfile <<~L
+        GEM
+          remote: #{file_uri_for(gem_repo2)}/
+          specs:
+            rack (1.0.0)
+
+        CHECKSUMS
+          #{expected_checksums}
+      L
+
       bundle :install, :raise_on_error => false
-      expect(exitstatus).to eq(37)
       expect(err).to include("Bundler found mismatched checksums.")
       expect(err).to include("1. remove the gem at #{cached_gem("rack-1.0.0")}")
+      expect(exitstatus).to eq(37)
 
       expect(cached_gem("rack-1.0.0")).to exist
       cached_gem("rack-1.0.0").rmtree
```

Conversely, if I add the lockfile that you added here to the spec in current master branch, it no longer raises the error.

Reply:
Thanks for figuring this out. It makes sense that it needs the checksum stored in the lockfile to trigger this. It probably was inserting the checksum automatically when they were enabled by default so I didn't realize it was necessary for the feature to work.